luigi/contrib/s3.py

```diff
@@ -319,61 +317,59 @@ def copy(self, source_path, destination_path, threads=100, start_time=None, end_
-        total_keys = 0
-        total_size_bytes = 0
 
         if self.isdir(source_path):
-            copy_jobs = []
-            management_pool = ThreadPool(processes=threads)
-
-            (bucket, key) = self._path_to_bucket_and_key(source_path)
-            key_path = self._add_path_delimiter(key)
-            key_path_len = len(key_path)
-            src_prefix = self._add_path_delimiter(src_key)
-            dst_prefix = self._add_path_delimiter(dst_key)
-
-            for item in self.list(source_path, start_time=start_time, end_time=end_time, return_key=True):
-                path = item.key[key_path_len:]
-                # prevents copy attempt of empty key in folder
-                if path != '' and path != '/':
-                    total_keys += 1
-                    total_size_bytes += item.size
-                    copy_source = {
-                        'Bucket': src_bucket,
-                        'Key': src_prefix + path
-                    }
-
-                    the_kwargs = {'Config': transfer_config, 'ExtraArgs': kwargs}
-                    job = management_pool.apply_async(self.s3.meta.client.copy,
-                                                      args=(copy_source, dst_bucket, dst_prefix + path),
-                                                      kwds=the_kwargs)
-                    copy_jobs.append(job)
-
-            # Wait for the pools to finish scheduling all the copies
-            management_pool.close()
-            management_pool.join()
-
-            # Raise any errors encountered in any of the copy processes
-            for result in copy_jobs:
-                result.get()
-
-            end = datetime.datetime.now()
-            duration = end - start
-            logger.info('%s : Complete : %s total keys copied in %s' %
-                        (datetime.datetime.now(), total_keys, duration))
+            return self._copy_dir(src_bucket, src_key, dst_bucket, dst_key, source_path,
+                                  transfer_config, threads=100, start_time=start_time, end_time=end_time, **kwargs)
 
         # If the file isn't a directory just perform a simple copy
         else:
-            total_keys += 1
-            copy_source = {
-                'Bucket': src_bucket,
-                'Key': src_key
-            }
             item = self.get_key(source_path)
-            total_size_bytes += item.size
-            self.s3.meta.client.copy(
-                copy_source, dst_bucket, dst_key, Config=transfer_config, ExtraArgs=kwargs)
+            self._copy_file(src_bucket, src_key, dst_bucket, dst_key, transfer_config, **kwargs)
+            return 1, item.size
+
+    def _copy_file(self, src_bucket, src_key, dst_bucket, dst_key, transfer_config, **kwargs):
+        copy_source = {
+            'Bucket': src_bucket,
+            'Key': src_key
+        }
+
+        self.s3.meta.client.copy(copy_source, dst_bucket, dst_key, Config=transfer_config, ExtraArgs=kwargs)
```

Comment:
It is still weird to be that `_copy_dir()` and `_copy_file()` don't return the same thing. `_copy_dir()` returns a tuple and `_copy_file()` returns `None`

Reply:
I wonder if it's really more efficient to use boto's `meta.client.copy()` for a single file rather than treating it with the same logic (minus the iteration over a dir) as `_copy_dir()`

Reply:
Regarding returning values I agree, it should be consistent.
Regarding efficiency, I think calling `meta.client.copy` could be more efficient and at the same time less consistent regarding Luigi S3 `_copy_file` logic. But since now we don't do much there, I agree on using the boto3 function. Good catch.  
What do you think?